tests/unit/s2n_resume_test.c

```diff
@@ -1618,5 +1618,169 @@ int main(int argc, char **argv)
+                EXPECT_NOT_NULL(conn);
+                EXPECT_SUCCESS(s2n_connection_set_config(conn, config));
+
+                /* Trigger the deserialization failure as late as possible.
+                 * The last byte is optional, so drop the last two bytes.
+                 */
+                size_t bad_ticket_size = ticket_size - 2;
+
+                /* Test the connection is not modified by a failed deserialize */
+                uint8_t saved_conn[sizeof(struct s2n_connection)] = { 0 };
+                EXPECT_MEMCPY_SUCCESS(saved_conn, conn, sizeof(struct s2n_connection));
+                uint8_t saved_secure[sizeof(struct s2n_crypto_parameters)] = { 0 };
+                EXPECT_MEMCPY_SUCCESS(saved_secure, conn->secure, sizeof(struct s2n_crypto_parameters));
+                EXPECT_FAILURE_WITH_ERRNO(s2n_connection_set_session(conn, ticket_bytes, bad_ticket_size),
+                        S2N_ERR_STUFFER_OUT_OF_DATA);
+                EXPECT_BYTEARRAY_EQUAL(saved_conn, conn, sizeof(struct s2n_connection));
+                EXPECT_BYTEARRAY_EQUAL(saved_secure, conn->secure, sizeof(struct s2n_crypto_parameters));
+
+                /* No valid ticket */
+                EXPECT_EQUAL(s2n_connection_get_session_length(conn), 0);
+            };
+
+            /* Test that deserialize does not modify the connection on a cipher selection failure,
+             * given the constraints proven above.
+             */
+            {
+                DEFER_CLEANUP(struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT),
+                        s2n_connection_ptr_free);
+                EXPECT_NOT_NULL(conn);
+                EXPECT_SUCCESS(s2n_connection_set_config(conn, config));
+
+                /* Trigger the deserialization failure when checking the validity
+                 * of the chosen cipher, not when parsing the ticket.
+                 */
+                EXPECT_SUCCESS(s2n_connection_set_cipher_preferences(conn, "null"));
+
+                /* Test the connection is not modified by a failed deserialize */
+                uint8_t saved_conn[sizeof(struct s2n_connection)] = { 0 };
+                EXPECT_MEMCPY_SUCCESS(saved_conn, conn, sizeof(struct s2n_connection));
+                uint8_t saved_secure[sizeof(struct s2n_crypto_parameters)] = { 0 };
+                EXPECT_MEMCPY_SUCCESS(saved_secure, conn->secure, sizeof(struct s2n_crypto_parameters));
+                EXPECT_FAILURE_WITH_ERRNO(s2n_connection_set_session(conn, ticket_bytes, ticket_size),
+                        S2N_ERR_CIPHER_NOT_SUPPORTED);
+                EXPECT_BYTEARRAY_EQUAL(saved_conn, conn, sizeof(struct s2n_connection));
+                EXPECT_BYTEARRAY_EQUAL(saved_secure, conn->secure, sizeof(struct s2n_crypto_parameters));
+
+                /* No valid ticket */
+                EXPECT_EQUAL(s2n_connection_get_session_length(conn), 0);
+            };
+        };
+
+        /* Invalid TLS1.3 tickets should not modify the connection.
```

Comment:
Do you think it's worth updating the usage guide or the set_session() documentation to mention that invalid tickets are a recoverable error? Otherwise they might not know that they don't have to recreate the connection if this function fails.